### PR TITLE
In-Memory Cache for CF Resources (Service Instances)

### PR DESCRIPTION
**Pull request by santiago-ventura**
## Summary
We want to introduce a mechanism for batching and caching CF resources. By storing the CF resources in memory, we aim to reduce the number of requests to the CF API and avoid reaching the rate limit.

##  Approach

- The cache feature is optional and can be enabled via the environment variable `RESOURCE_CACHE_ENABLED`, which can have values of `true` or `false`.

- Another environment variable, `CACHE_TIMEOUT`, is introduced to control the refresh of the cached resources, based on a `lastCacheTime` attribute of the Cache instance. 

- At the start of the operator, an instance of a Config object is created. This config object has two attributes with values of the environment variables described above.

- The first time the' spaceClient' is created, an instance of the Cache object is created and populated. 

-  The Cache instance will be refreshed every time the cache time-out is reached.

> - This PR has been tested manually in one of our Kubernetes Clusters.
> - A second PR will be following this one; implementing the solution for the CF resources Space, and ServiceBinding. A third PR will have the documentation and the integration, and unit test.

### CODE REVIEW

File: internal/cf/client.go [hunk 1]
```diff
@@ -51,8 +56,10 @@ type clientCacheEntry struct {
 }
 
 var (
-	cacheMutex  = &sync.Mutex{}
-	clientCache = make(map[clientIdentifier]*clientCacheEntry)
+	cacheMutex        = &sync.Mutex{}
+	clientCache       = make(map[clientIdentifier]*clientCacheEntry)
+	cfCache           *facade.Cache
```
Comment (RalfHammer) [inline internal/cf/client.go:61]: use term "resourceCache" to have a clearer definition (and a better separation from clientCache)

File: internal/cf/client.go [hunk 2]
```diff
@@ -51,8 +56,10 @@ type clientCacheEntry struct {
...
+	refreshCacheMutex = &sync.Mutex{}
```
Comment (RalfHammer) [inline internal/cf/client.go:62]: use term "resourceCacheMutex" to link the name to the name of the cache

File: internal/cf/client.go [hunk 3]
```diff
@@ -51,8 +56,10 @@ type clientCacheEntry struct {
 }
 
 var (
-	cacheMutex  = &sync.Mutex{}
-	clientCache = make(map[clientIdentifier]*clientCacheEntry)
+	cacheMutex        = &sync.Mutex{}
```
Comment (RalfHammer) [inline internal/cf/client.go:59]: With your changes, we now have to caches.
To make it clearer, use more specific terms like "clientCacheMutex" also for this variable

File: internal/cf/client.go [hunk 4]
```diff
@@ -114,7 +127,10 @@ func newSpaceClient(spaceGuid string, url string, username string, password stri
 	if err != nil {
 		return nil, err
 	}
-	return &spaceClient{spaceGuid: spaceGuid, client: *c}, nil
+
+	spcClient := &spaceClient{spaceGuid: spaceGuid, client: *c}
```
Comment (RalfHammer) [inline internal/cf/client.go:131]: spc might be a misleading abbreviation.
Maybe you could use cfSpaceClient?

File: internal/cf/client.go [hunk 5]
```diff
@@ -158,10 +174,10 @@ func NewSpaceClient(spaceGuid string, url string, username string, password stri
 	cacheEntry, isInCache := clientCache[identifier]
 
 	var err error = nil
-	var client *spaceClient = nil
+	var spcClient *spaceClient = nil
```
Comment (RalfHammer) [inline internal/cf/client.go:177]: spc is a misleading abbreviation, see above

File: internal/config/config.go [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+
+import (
+	"github.com/caarlos0/env/v11"
+)
+
+// Config defines the configuration keys
+type Config struct {
+
+	//Resource cache is enabled or disabled
+	IsResourceCacheEnabled bool `env:"RESOURCE_CACHE_ENABLED" envDefault:"true"`
```
Comment (RalfHammer) [inline internal/config/config.go:16]: By default, we want to have the "old" behaviour without the cache.
Therefore, you should set the default to false.

File: internal/cf/client_test.go [hunk 7]
```diff
@@ -49,6 +50,15 @@ func TestCFClient(t *testing.T) {
...
 
+// is resource cache enabled and cache timeout
+var resourceCacheEnabled = false
+var resourceCacheTimeout = 5 * time.Minute
+
+var cfg = &config.Config{
+	IsResourceCacheEnabled: resourceCacheEnabled,
+	CacheTimeOut:           resourceCacheTimeout.String(),
+}
+
```
Comment (RalfHammer) [inline internal/cf/client_test.go:61]: Is this duplicate from suite_test.go?

File: internal/facade/client.go [hunk 8]
```diff
@@ -88,6 +99,135 @@ type SpaceClient interface {
...
+	// AddInstanceInCache(key string, instance *Instance)
+	// GetInstanceFromCache(key string) (*Instance, bool)
+	// AddBindingInCache(key string, binding *Binding)
+	// GetBindingFromCache(key string) (*Binding, bool)
+}
+
+type SpaceClientBuilder func(string, string, string, string, config.Config) (SpaceClient, error)
+
+// Cache is a simple in-memory cache to store spaces, instances, and bindings
+type Cache struct {
```
Comment (RalfHammer) [inline internal/facade/client.go:113]: use term "ResourceCache" to be more specific

File: internal/facade/client.go [hunk 9]
```diff
@@ -88,6 +99,135 @@ type SpaceClient interface {
...
+	expirationTime := c.lastCacheTime.Add(c.cacheTimeOut)
+	//expiryTime := time.Until(c.lastCacheTime)
+	fmt.Printf("Expiry time: %v\n", expirationTime)
+	fmt.Printf("Cache timeout: %v\n", c.cacheTimeOut)
+	return time.Now().After(expirationTime)
+
+}
+func (c *Cache) SetLastCacheTime() {
+	c.lastCacheTime = time.Now()
+	fmt.Printf("Last cache time: %v\n", c.lastCacheTime)
```
Comment (RalfHammer) [inline internal/facade/client.go:187]: Either remove this or use debug level of logging API

File: internal/facade/client.go [hunk 10]
```diff
@@ -88,6 +99,135 @@ type SpaceClient interface {
...
+
+// Function to set the Last cache time
+func (c *Cache) GetLastCacheTime() time.Time {
+	return c.lastCacheTime
+}
+
+func (c *Cache) IsCacheExpired() bool {
+
+	expirationTime := c.lastCacheTime.Add(c.cacheTimeOut)
+	//expiryTime := time.Until(c.lastCacheTime)
```
Comment (RalfHammer) [inline internal/facade/client.go:179]: Remove this

File: internal/facade/client.go [hunk 11]
```diff
@@ -88,6 +99,135 @@ type SpaceClient interface {
...
+type SpaceClientBuilder func(string, string, string, string, config.Config) (SpaceClient, error)
+
+// Cache is a simple in-memory cache to store spaces, instances, and bindings
+type Cache struct {
+	spaces                 map[string]*Space
+	instances              map[string]*Instance
+	bindings               map[string]*Binding
+	mutex                  sync.RWMutex
+	lastCacheTime          time.Time
+	cacheTimeOut           time.Duration
```
Comment (RalfHammer) [inline internal/facade/client.go:119]: I am not sure if we might want to have different timeouts depending on the resource type

File: main.go [hunk 12]
```diff
@@ -98,7 +100,11 @@ func main() {
 		setupLog.Error(err, "unable to parse webhook bind address", "controller", "Space")
 		os.Exit(1)
 	}
-
+	cfg, err := config.Load()
+	if err != nil {
+		fmt.Printf("failed to load config %v\n", err)
```
Comment (RalfHammer) [inline main.go:105]: use setupLog instead of Printf

File: internal/cf/client.go [hunk 13]
```diff
@@ -31,11 +34,13 @@ const (
...
 	client           cfclient.Client
+	//resourcesCache   *facade.Cache
 }
 
 type spaceClient struct {
-	spaceGuid string
-	client    cfclient.Client
+	spaceGuid      string
+	client         cfclient.Client
+	resourcesCache *facade.Cache
```
Comment (RalfHammer) [inline internal/cf/client.go:43]: I think the singular form "resourceCache" is already sufficient

File: internal/cf/client.go [hunk 14]
```diff
@@ -83,6 +90,12 @@ func newOrganizationClient(organizationName string, url string, username string,
 	if err != nil {
 		return nil, err
 	}
+	// isResourceCacheEnabled, _ := strconv.ParseBool(os.Getenv("RESOURCE_CACHE_ENABLED"))
```
Comment (RalfHammer) [inline internal/cf/client.go:93]: You mentioned that this is part of another PR.
Therefore, just remove it from here?

File: internal/cf/client.go [hunk 15]
```diff
@@ -212,3 +236,46 @@ func NewSpaceHealthChecker(spaceGuid string, url string, username string, passwo
 
 	return client, err
 }
+
+func (c *spaceClient) populateResourcesCache() {
```
Comment (RalfHammer) [inline internal/cf/client.go:240]: singular from "populateResourceCache" is sufficient, see above

File: internal/facade/client.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (santiago-ventura) [reply internal/facade/client.go:187]: We will clean the code once we finish the manual test. This particular go file does not use logs, so want to keep it as it is in the main branch.

**Pull request closed without merging**